src/Juvix/Compiler/Core/Extra/Recursors/RMap.hs

```diff
@@ -0,0 +1,111 @@
+module Juvix.Compiler.Core.Extra.Recursors.RMap where
+
+import Data.Functor.Identity
+import Juvix.Compiler.Core.Data.BinderList qualified as BL
+import Juvix.Compiler.Core.Extra.Base
+import Juvix.Compiler.Core.Extra.Recursors.Base
+import Juvix.Compiler.Core.Extra.Recursors.Utils
+
+data BinderRemove = BinderRemove
+  { _binderRemoveBinder :: Binder,
+    _binderRemoveNode :: Node
+  }
+
+makeLenses ''BinderRemove
+
+data BinderChange
+  = -- | `BCAdd n` -- add `n` binders
+    BCAdd Int
+  | -- | `BCKeep b` -- keep the binder `b`
+    BCKeep Binder
+  | -- | `BCRemove (BinderRemove b n)` -- remove the binder `b` and replace the
+    -- occurrences of the bound variable with the node `n`; the de Bruijn
+    -- indices of `n` are with respect to the result
+    BCRemove BinderRemove
+
+bindersFromBinderChange :: [BinderChange] -> [Binder]
+bindersFromBinderChange =
+  reverse
+    . foldl'
+      ( \bs chg -> case chg of
+          BCAdd _ -> bs
+          BCKeep b -> b : bs
+          BCRemove (BinderRemove b _) -> b : bs
+      )
+      []
+
+bindersNumFromBinderChange :: [BinderChange] -> Int
+bindersNumFromBinderChange =
```

Comment:
It is a bit unintuitive to me that the number of binders in `BCAdd` is ignored. Could it be clarified in a comment why we want that?

Reply:
Because these are the binders in the *original* node, not in the result. I'll add a comment. I'll also add similar helper functions for the binders in the result node.

Reply:
Done